controllers/openstackmachine_controller.go

```diff
@@ -292,11 +302,155 @@ func (r *OpenStackMachineReconciler) reconcileDelete(scope scope.Scope, cluster
 		return ctrl.Result{}, fmt.Errorf("delete instance: %w", err)
 	}
 
+	if err := r.reconcileDeleteFloatingAddressFromPool(scope, openStackMachine); err != nil {
+		return ctrl.Result{}, err
+	}
+
 	controllerutil.RemoveFinalizer(openStackMachine, infrav1.MachineFinalizer)
 	scope.Logger().Info("Reconciled Machine delete successfully")
 	return ctrl.Result{}, nil
 }
 
+func (r *OpenStackMachineReconciler) reconcileFloatingIPAddressFromPool(ctx context.Context, scope scope.Scope, openStackMachine *infrav1.OpenStackMachine, openStackCluster *infrav1.OpenStackCluster) (*ipamv1.IPAddressClaim, *ctrl.Result, error) {
+	if openStackMachine.Spec.FloatingAddressFromPool == nil {
+		return nil, nil, nil
+	}
+	var claim *ipamv1.IPAddressClaim
+	claim, err := r.getOrCreateIPAddressClaimForFloatingAddress(ctx, scope, openStackMachine, openStackCluster)
+	if err != nil {
+		conditions.MarkFalse(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition, infrav1.FloatingAddressFromPoolErrorReason, clusterv1.ConditionSeverityError, "Failed to reconcile floating IP claims: %v", err)
+		return nil, &ctrl.Result{}, err
+	}
+	if claim.Status.AddressRef.Name == "" {
+		return claim, &ctrl.Result{RequeueAfter: waitForIPAddressClaimToBeAllocated}, nil
+	}
+	return claim, nil, nil
+}
+
+// createIPAddressClaim creates IPAddressClaim for the FloatingAddressFromPool if it does not exist yet.
+func (r *OpenStackMachineReconciler) getOrCreateIPAddressClaimForFloatingAddress(ctx context.Context, scope scope.Scope, openStackMachine *infrav1.OpenStackMachine, openStackCluster *infrav1.OpenStackCluster) (*ipamv1.IPAddressClaim, error) {
+	poolRef := openStackMachine.Spec.FloatingAddressFromPool
+	claimName := names.GetFloatingAddressIPAddressClaimName(openStackMachine.Name)
+	claim := &ipamv1.IPAddressClaim{}
+
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: openStackMachine.Namespace, Name: claimName}, claim); err != nil {
+		if apierrors.IsNotFound(err) {
+			claim = &ipamv1.IPAddressClaim{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      claimName,
+					Namespace: openStackMachine.Namespace,
+					Labels: map[string]string{
+						clusterv1.ClusterNameLabel: openStackCluster.Labels[clusterv1.ClusterNameLabel],
+					},
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: openStackMachine.APIVersion,
+							Kind:       openStackMachine.Kind,
+							Name:       openStackMachine.Name,
+							UID:        openStackMachine.UID,
+						},
+					},
+				},
+				Spec: ipamv1.IPAddressClaimSpec{
+					PoolRef: *poolRef,
+				},
+			}
+			if err := r.Client.Create(ctx, claim); err != nil {
+				conditions.MarkFalse(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition, infrav1.FloatingAddressFromPoolErrorReason, clusterv1.ConditionSeverityError, "Failed to reconcile floating IP claims: %v", err)
```

Comment:
Note that when we return here the caller is going to set this false condition again, overwriting this one.

Please can we move all error condition setting out of this function into the caller?